### PR TITLE
Respect the ScmComunicationException on token validation step

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
* Throw the `ScmCommunicationException` on token validation step instead of returning invalid result.
* Propagate the `ScmCommunicationException` to the dashboard as an `ApiException`.
* Throw `ScmUnauthorizedException` on each scm api request with invalid token.
* Log error instead of throwing `RuntimeException` on workspace provision steps.

### What issues does this PR fix or reference?

https://github.com/eclipse-che/che/issues/23076

### How to test this PR?

1. Deploy Che from the PR image: `quay.io/ivinokur/che-server:che-23076` The image is built with a modification: to simulate the unreachable SCM server I changed the api endpoint url in the GitHub api client:
https://github.com/eclipse-che/che-server/blob/e04d787ed847f06faf9d18c56f0774de3cfa8b7c/wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java#L57
2. Configure Che to use dashboard image from the related pull request: `quay.io/eclipse/che-dashboard:pr-1179`
3. Configure [GitHub oauth](https://eclipse.dev/che/docs/stable/administration-guide/configuring-oauth-2-for-github/)
4. Start a workspace from a GitHub repository with devfile.

See: a warning notification appears during workspace startup:
Screenshot 2024-08-30 at 13 59 48
The workspace starts successfully. 

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

**Comment by artaleks9** [comment]
By some reason, the changes in test file are not staged.
Maybe there is changed UI, or something like that.
I investigate it.
pr-715

**Comment by artaleks9** [comment]
@vinokurig 
The `ci/prow/v14-gitlab-with-oauth-setup-flow` is fail due to changes in IDE
I provided the screenshots to compare and I'm going to fix up it in our Che test repository
The functionality related to git provider OAuth works properly. I've checked it manually using `quay.io/eclipse/che-server:pr-715`
pr-715-che-IDE
pr-715-no-change-IDE

**Comment by artaleks9** [comment]
So, there are no restrictions related tests to merge PR.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/CredentialsSecretConfigurator.java [hunk 1]
```diff
@@ -79,7 +83,7 @@ public void configure(NamespaceResolutionContext namespaceResolutionContext, Str
                   | ScmConfigurationPersistenceException
                   | UnsatisfiedScmPreconditionException
                   | ScmUnauthorizedException e) {
-                throw new RuntimeException(e);
+                LOG.error(e.getMessage(), e);
```
Comment (tolusha) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/CredentialsSecretConfigurator.java:86]: What is the benefits of logging error?

File: wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketApiClient.java [hunk 2]
```diff
@@ -228,6 +235,8 @@ private <T> T executeRequest(
             throw new ScmBadRequestException(body);
           case HTTP_NOT_FOUND:
             throw new ScmItemNotFoundException(body);
+          case HTTP_UNAUTHORIZED:
+            throw new ScmUnauthorizedException(body, "github", "v1", "");
```
Comment (tolusha) [inline wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketApiClient.java:239]: Can we move constant somewhere else? 
Copy/paste from github?

File: wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java [hunk 3]
```diff
@@ -290,6 +296,8 @@ private <T> T executeRequest(
             throw new ScmBadRequestException(body);
           case HTTP_NOT_FOUND:
             throw new ScmItemNotFoundException(body);
+          case HTTP_UNAUTHORIZED:
+            throw new ScmUnauthorizedException(body, "github", "v1", "");
```
Comment (tolusha) [inline wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java:300]: Why v1 ?

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabApiClient.java [hunk 4]
```diff
@@ -159,6 +163,8 @@ private <T> T executeRequest(
             throw new ScmBadRequestException(body);
           case HTTP_NOT_FOUND:
             throw new ScmItemNotFoundException(body);
+          case HTTP_UNAUTHORIZED:
+            throw new ScmUnauthorizedException(body, "github", "v1", "");
```
Comment (tolusha) [inline wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabApiClient.java:167]: copy/paste from github ?

**Review by ibuziuk — COMMENTED** [review]
@vinokurig great job, but can we make UX more informative - Error message should say that Github might be down + add a link to the statuspage - https://www.githubstatus.com/

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/CredentialsSecretConfigurator.java (hunk 1 above)
Comment (vinokurig) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/configurator/CredentialsSecretConfigurator.java:86]: We do not interrupt the workspace start in this case.

File: wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketApiClient.java (hunk 2 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory-bitbucket/src/main/java/org/eclipse/che/api/factory/server/bitbucket/BitbucketApiClient.java]: >Copy/paste from github?

right, fixed
>Can we move constant somewhere else?

Could you please elaborate on this?

File: wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java (hunk 3 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/GithubApiClient.java]: Changed to `v2` corresponds to oauth version.

File: wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabApiClient.java (hunk 4 above)
Comment (vinokurig) [reply wsmaster/che-core-api-factory-gitlab/src/main/java/org/eclipse/che/api/factory/server/gitlab/GitlabApiClient.java]: fixed

**Review by tolusha — APPROVED** [review]

**Comment by vinokurig** [comment]
@ibuziuk 
> can we make UX more informative - Error message should say that Github might be down + add a link to the statuspage - https://www.githubstatus.com/

I've improved the notification text but unfortunately there is no quick way to have the link clickable in the warning notification, would that be ok in the scope of the issue? cc @olexii4 @akurinnoy 
screenshot-eclipse-che apps rosa z9hnf-8cdw2-7m6 qpi4 p3 openshiftapps com-2024 09 11-15_49_13

**Comment by ibuziuk** [comment]
We should either make the link clickable or just reword the warning without the link - `GitHub might not be operational. Check the status page.`

**Comment by vinokurig** [comment]
@ibuziuk I have removed the links from the notification and opened an [issue to be able to add links](https://github.com/eclipse-che/che/issues/23142)
screenshot-eclipse-che apps rosa x9djd-x3aqj-zvk qktx p3 openshiftapps com-2024 09 12-15_59_52

**Review by ibuziuk — APPROVED** [review]

**Review by artaleks9 — APPROVED** [review]

**Pull request merged**